Balance endpoint - to be retired

paweljakubas opened the pull request:
# Issue Number

ADP-656

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] added endpoint, all Api types, swagger
- [x] Api types tested in unit testing and passing
- [x] added Server's balanceTransaction function capturing the logic
- [x] extended Transaction interface : updateTx
- [x] added deserialisation tests in shelley unit tests
- [x] added error handling
- [x] did some very first version coin selection with external inputs, to be replaced with ADP-1070
- [x] tested in integration testing several errors (already covered tx, wrong format, cannot deserialise) plus successful rebalance - with wallet's generated CBORs 
- [x] implement updateSealedTx in shelley being impl of interface updateTx
- [x] add testing for deserialisation/coin selection for plutus examples


# Comments

Based on the PR #2896 branch.

paweljakubas committed:
Implement balanceTransaction

Post-rebase fixes

work on json instances and align code after rebase

add signatories to ApiBalancedTransactionPostData

work with core unit testing

correct Maformed and work on passing core unit test suite

add two integration test with realistic cborHexes, one fully covered, one with missing inputs

more detail added in Server.balanceTransaction

some cleanings

the same binary expectation in first integration test

some cleanings

fake coin selection to verify that binary is returned in no more inputs case

better treating format and deserialization errors

add shelley core text envelop testing

add ErrBalanceTxAllOutputsCovered

add temporary workaround for coming coin selection with initial inputs

add updateTx function to Transaction interface

CompatibilitySpec

add first infrastructure for testing plutus examples

add other plutus examples to testing

use updateTx and constructTransaction properly in balanceTransaction

make sure core unit testing pass

add plutus cborHex example

better formatting

correct tokenNames to comply with what cardano-api produces

updateSealedTx impl - part 1

updateSealedTx impl - part 2

add plutus example to integration testing

Rebase fixes

jonathanknowles commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -187,20 +236,103 @@ withdrawalToCoin = \case
+-- derived from the user's spending passphrasee. This data type pairs an
+-- encrypted signing key with the passphrase needed to decrypt it.
+--
+-- We call this type "decrypted", because although the key is still encrypted, the passphrase is right there, so it may as well be decrypted.
```
```suggestion
-- We call this type "decrypted", because although the key is still encrypted,
-- the passphrase is right there, so it may as well be decrypted.
```

jonathanknowles commented on lib/test-utils/src/Test/Utils/Pretty.hs:
```diff
@@ -39,3 +40,8 @@ instance Show a => ToText (Pretty a)
 pShowBuilder :: Show a => a -> Builder
 pShowBuilder = fromLazyText . pShow
+
+-- | Like '===, but prettier.
```
```suggestion
-- | Like '===', but prettier.
```

paweljakubas committed:
Regenerate nix

rvl committed:
Change SealedTx to have dual representations

This type now contains both a serialised and deserialised
representation of the transaction.

This also adds a stub for the signTx servant handler.

sevanspowell commented on specifications/api/swagger.yaml:
```diff
@@ -2755,6 +2858,47 @@ components:
+        txOut: *txOutNode
+
+    ApiExternalInputs: &ApiExternalInputs
+      description: A flat list of node compliant inputs.
```
I think I'm missing something, can't we just get the inputs from the serialised transaction itself?

sevanspowell commented on specifications/api/swagger.yaml:
```diff
@@ -2755,6 +2858,47 @@ components:
             - <<: *transactionWithdrawalRequestSelf
             - <<: *transactionRedemptionRequest
 
+    ApiTransactionExtended: &ApiTransactionExtended
```
Could we perhaps describe what we expect this transaction to look like? I can tell from this description that it's CBOR encoded, but what might the transaction look like before it's CBOR encoded? Do we expect it to be missing inputs and/or outputs?

sevanspowell replied:
Just as someone with no context, I'm not sure how I might generate a transaction to call this endpoint.

sevanspowell commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2016,6 +2023,85 @@ postTransactionFeeOld ctx (ApiT wid) body = do
+        ( map toTxIn $ NE.toList $ cs ^. #inputs
+        , map toTxOut $ cs ^. #change)
+
+    areOutputsCovered (Tx _ feeM _ outs _ _) = -- TODO in ADP-656 is it correct- double check
```
I think there's value in moving meaningful logic outside the handler. For example, this concept of "Is this transaction balanced" could be useful to other modules, and is more easily tested outside the handler. Inside the handler, it's very difficult to test this logic.

paweljakubas committed:
swagger: Add balanceTx

paweljakubas committed:
Api.TypesSpec: Add golden JSON files

paweljakubas committed:
Add example plutus transactions

Anviking commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -621,6 +622,9 @@ errMsg400ScriptNotUniformRoles :: String
 
+errMsg403OutputsAlreadyCovered :: String
+errMsg403OutputsAlreadyCovered =
+    "The transaction to be balanced has already all outputs covered. Please send a transaction that requires more inputs to be picked to be balanced."
```
If PAB _happens_ to construct a balanced tx, shouldn't the balance endpoint be a no-op rather than failing?

> Please send a transaction that requires more inputs to be picked to be balanced.

"Please find a way to make the tx more expensive" sounds like an odd recommendation...

paweljakubas replied:
yes, message is unfortunate for two reasons: first of all, unbalanced tx is the one that lacks outputs OR inputs, Secondly, we should rather write that sent tx is balanced and does not need further balancing but signing and then submitting.

paweljakubas replied:
I resigned from this error naming in https://github.com/input-output-hk/cardano-wallet/pull/2900

Anviking replied:
I was more wondering: why not remove the error?

If the endpoint succeeds in the — I take rare event of PAB submitting an already balanced tx, then they wouldn't have to add special handling for:
> we should rather write that sent tx is balanced and does not need further balancing but signing and then submitting.

Anviking replied:
Also seemed to me that it _might_ be easier for _us too_, to not care about defining `areOutputsCovered`. I would presume the code responsible for balancing the tx would already be able to notice when the tx was balanced, and return without any changes… but could be wrong.

paweljakubas replied:
The problem is that I am reusing ApiCoinSelection in the response of endpoint and it cannot be empty.So in case of already balance tx I tentatively choose to use some error to circumvent this problem. As a first step.

rvl committed:
Rework calcScriptExecutionCost using SealedTx

rvl committed:
Make mockSealedTx work better

rvl committed:
HLint
